### ISSUE TITLE
Delay application of experience gain to limit/forbid grinding

### ISSUE THREAD

**Issue by keyspace** [CleverRaven/Cataclysm-DDA#20203]
Mentioned in an [OT comment in PR #20165](https://github.com/CleverRaven/Cataclysm-DDA/pull/20165#issuecomment-277536166):

> Grinding is possible due to current mechanic of experience gain: it is applied immediately. Only solution I've come up with is: _not_ applying it immediately. Say, drop it into a "buffer" instead, and limit stacking while doing so. Process the "buffer" as time goes by. "You think over your recent actions. Your survival skill has increased to 1!"

Previous discussions (of a now-outdated mechanic):

* Rethinking XP and skill comprehension #613
* Skill-raising: how to fix it without killing realism and without breaking the game #5250
* Limiting skill gain for crafting skills #6465

Issue outlined in the latter two has been addressed: one can no longer craft clean water to get high cooking, etc..

What I propose here, though, is moving application of `SkillLevel::train()` from `player::practice()` to an external timed processor.

E.g., invoking `player::practice()` checks-and-writes to a "clearing house" structure. Yet-unimplemented magic checks this structure once in X, and does the training.

Please do discuss:

* [ ] whether it has been discussed/attempted before;
* [x] whether this would truly prevent grinding, and what further exploits could arise;
* [ ] the general viability of such a system;
* [ ] if viable:
  * [ ] whether it should be a game option;
  * [ ] when and how often should practice->train conversion be done;
  * [ ] whether it could/should be dependent on intelligence/morale/pain/fatigue;
  * [ ] the impact it could have if skill rust is enabled, or other game options gaining more influence;
  * [ ] the impact it could have on book reading;
  * [ ] how this will/should impact `isTraining()` and `toggleTraining()` display in UI;
  * [ ] other, not listed below.

Please don't discuss:

* monster/mapgen loot drop balance;
* which activities should practice what skill;
* changing the skill list.

-----

EDIT: expanded lists above
EDITs: many, tracking of discussion

**Comment by keyspace** [comment]
The `isTraining()` mechanic is peculiar. Currently, all skills start out as that, and the only place that seems to be checked is in `player.cpp`, in `player::practice()`:

``` c++
11676     if (amount > 0 && level.isTraining()) {
11677         int oldLevel = get_skill_level( id );
11678         get_skill_level( id ).train(amount);
...
```

What is the purpose of this "toggling what to train"? `SAVANT` characters getting to choose?..

TBH, I just found out of the possibility to toggle training at all!

`git blame` fails to display its origin properly for me, and [github's blame](https://github.com/CleverRaven/Cataclysm-DDA/blame/master/src/player.cpp) is so humongous it sometimes fails to load. Its current blame shows to commit 5760dd004d7402802ef0945319b32d285b84cc60 (part of PR #7822 - Cap construction practice), and it's pure styling.

-----

`git blame src/skill.h` suggests that was added in commit f6e605a8edb85257272913c8bef9bcd57a5c8760 (part of PR #25 - Loadable skills), but that doesn't specify _why_ it was needed, and requires reviewing a huge PR for whether it's a reimplementation of a previous mechanic, or something new. :/

**Comment by kevingranade** [comment]
Basically you want to throttle how often practice succeeds.  I don't follow how this improves anything, you're basically required to try and fit an action of each kind you're interested in into each time period, so instead of "do mechanics stuff a bunch", which is at least intuitive, you want to either take a break once every X turns to do something "mechanics-y", or pick some number of activities for skills you're interested in and rotate through them.

Regarding
> whether this would truly prevent grinding

I don't think it would, it would just require someone trying to raise skills quickly to follow a spaced repition training regimen.

Re: "toggling training" Your 2nd guess is correct, it allows you to perform actions without burning focus on them.  Yes it's kind of a hack, but a necessary one as long as ordinary activities drain focus.

As long as you're reviewing old ideas for skill overhauls, throw this one in: http://smf.cataclysmdda.com/index.php?topic=3728.0
tl;dr (but do read it :P) it "embraces grinding" by adding long "practice actions" players can take to raise skills over time. The idea being that like crafting, you accumulate materials and unlock recipes in order to spend time and energy in exchange for skillups.

**Comment by keyspace** [comment]
> to follow a spaced [repetition] training regimen.

Thought of this a bit. I find it likely that many do like a mechanic like this being non-complicated (and therefore intuitive), and don't mind much the "grinding" aspect.

When I first had this idea a few months ago, thought of experience applying during sleep. But later discarded this idea, having played a character that couldn't sleep that much, for various reasons.

> tl;dr (but do read it :P)

Oh, I did! Thanks for the link!.. Will dwell on it.

-----

I've started to play with this, strictly "what if", and will link a branch if something viable comes of it. Would expect discussion to last a while (say, a few months), so won't be opening a [WIP] PR.

**Comment by 1skandar** [comment]
Ya know, instead of trying to get rid of grinding maybe it would be better to address *why* grinding is necessary. Often, there is no non-grind way to raise a skill the few times. You simply are not going to raise tailoring to 1 or 2 without grinding, nor mechanics. Or a lot of other skills. There is simply no way you'd naturally get those first few levels just going about playing the game.

Now, look instead at survival. You want to gain the first point in survival? You go foraging in the underbrush. Is it a bit grindy? Well, maybe, but it also serves an in-game point, finding food. And beyond that you butcher corpses for meat. There is no need to grind survival because you can raise it to useful levels just playing the game as normal.

Cure the disease, not just the symptoms. Give us something to do with each skill that is useful, and the need for grinding will naturally stop. For instance, raise the experience gain at level 0 mechanics for picking locks. Say 10 or 12 locks equals level 1. And from there adjust the skill level needed to work on vehicles and you can actually do useful stuff in game with that skill without ever grinding.

And whatever you do, make any non-grind thingie is optional. It's already punishing enough to try to get quite a few skills off the ground, making it harder is completely unnecessary.

**Comment by kevingranade** [comment]
Doesn't address the issue, (trades continuous grinding for intermittent but just as lengthy grinding).
